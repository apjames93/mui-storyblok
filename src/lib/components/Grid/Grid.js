import React from 'react';
import PropTypes from 'prop-types';
import MuiGrid from '@material-ui/core/Grid';
import Storyblok from '../../utils/Storyblok';
import sizeGrid from '../../utils/sizeGrid';
import { muiStringProp } from '../../utils/customProps';
import { renderComponents, storyBlokClickableProps } from '../../utils/customComponents';
import GridItem from './components/GridItem/GridItem';

const Grid = ({
  alignContent,
  alignItems,
  rootClass,
  direction,
  justify,
  wrap,
  spacing,
  content,
  components,
  style,
  _editable,
  _uid,
}) => {
  const styles = Storyblok.arrayToMuiStyles(rootClass, { padding: '25px', ...style });
  const gridRender = Object.keys(components);
  const clickable = storyBlokClickableProps({
    _editable,
    component: 'MuiGrid',
    _uid,
  });

  return (
    <MuiGrid
      container
      alignContent={alignContent}
      alignItems={alignItems}
      direction={direction}
      justify={justify}
      wrap={wrap}
      spacing={Number(spacing)}
      data-blok-c={clickable.dataBlokC}
      data-blok-uid={clickable.dataBlokUid}
      className={`${styles.root} ${clickable.storyblokClass}`}
    >
      {content.map((component, key) => {
        const sbClickable = storyBlokClickableProps(component);

        if (gridRender.includes(component.component)) {
          return renderComponents(components, { ...component, sizeGrid }, key);
        }

        return (
          <GridItem
            {...component}
            key={key}
            components={components}
            sizeGrid={sizeGrid}
            dataBlokC={sbClickable.dataBlokC}
            dataBlokUid={sbClickable.dataBlokUid}
            storyblokClass={sbClickable.storyblokClass}
          />
        );
      })}
    </MuiGrid>
  );
};

export default Grid;

Grid.propTypes = {
  /**
   * developer prop to pass components to a gridItem
   */
  components: PropTypes.shape().isRequired,
  /**
   * stroyblok multiselect of css classes
   * Mui Override or extend the styles applied to the component. See CSS API below for more details.
   */
  rootClass: PropTypes.arrayOf(PropTypes.string),
  /**
   * mui prop: 'stretch', 'center', 'flex-start', 'flex-end', 'space-between', 'space-around'
   * Defines the align-content style property. It's applied for all screen sizes.
   */
  alignContent(props, propName, componentName) {
    const validProp = ['stretch', 'center', 'flex-start', 'flex-end', 'space-between', 'space-around'];
    return muiStringProp(props, propName, componentName, validProp);
  },
  /**
   * mui prop: 'flex-start' , 'center' , 'flex-end' , 'stretch' , 'baseline'
   * Defines the align-items style property. It's applied for all screen sizes.
   */
  alignItems(props, propName, componentName) {
    const validProp = ['flex-start', 'center', 'flex-end', 'stretch', 'baseline'];
    return muiStringProp(props, propName, componentName, validProp);
  },
  /*
  * mui prop: 'row', 'row-reverse', 'column', 'column-reverse'
  * Defines the flex-direction style property. It is applied for all screen sizes.
  */
  direction(props, propName, componentName) {
    const validProp = ['row', 'row-reverse', 'column', 'column-reverse'];
    return muiStringProp(props, propName, componentName, validProp);
  },
  /**
   * mui prop:
   * 'flex-start', 'center', 'flex-end', 'space-between', 'space-around', 'space-evenly'
   * Defines the justify-content style property. It is applied for all screen sizes.
   */
  justify(props, propName, componentName) {
    const validProp = ['flex-start', 'center', 'flex-end', 'space-between', 'space-around', 'space-evenly'];
    return muiStringProp(props, propName, componentName, validProp);
  },
  /**
   * mui prop:  0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
   * Defines the space between the type item
   component. It can only be used on a type container component.
   */
  spacing: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * mui prop: 'nowrap', 'wrap', 'wrap-reverse'
   * Defines the flex-wrap style property. It's applied for all screen sizes.
   */
  wrap(props, propName, componentName) {
    const validProp = ['nowrap', 'wrap', 'wrap-reverse'];
    return muiStringProp(props, propName, componentName, validProp);
  },
  /**
   * Content passed to render
   */
  content: PropTypes.arrayOf(PropTypes.shape({
    component: PropTypes.string.isRequired,
  })).isRequired,

  /** style obj to pass to grid */
  style: PropTypes.shape(),
};

Grid.defaultProps = {
  alignContent: 'stretch',
  alignItems: 'center',
  rootClass: [],
  direction: 'row',
  justify: 'center',
  spacing: '2',
  wrap: 'wrap',
  style: {},
};
