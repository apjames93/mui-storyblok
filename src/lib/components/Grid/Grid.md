##### MuiGrid has the ability to render custom components.

MuiGrid available components:

* MuiGridItem
* MuiAppBar
* MuiTabs
* MuiMobileStepper
* MuiCircularProgress
* MuiGridList
* MuiHeroHeader
* MuiNotificationBanner