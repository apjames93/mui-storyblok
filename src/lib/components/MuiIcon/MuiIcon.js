import React, { createElement } from 'react';
import PropTypes from 'prop-types';
import Icon from '@material-ui/core/Icon';
import Storyblok from '../../utils/Storyblok';
import { muiStringProp } from '../../utils/customProps';
import BBBSvg from './components/BBBSvg/BBBSvg';
import FacebookSvg from './components/FacebookSvg/FacebookSvg';
import GoogleSvg from './components/GoogleSvg/GoogleSvg';
import LinkedinSvg from './components/LinkedinSvg/LinkedinSvg';
// import styles for icons in styleguide
import '../../styles/styles.scss';

// import icons to use for custom icons passed down to app;
import customIcons from '../../utils/customIcons';

// import { icons } from '../MuiStoryblok/MuiStoryblok';
// const icons = getIcons();
/**
 * MuiIcon is used in storyblok redirect to react routes
 */

export const MuiIcon = ({
  color,
  fontSize,
  iconName,
  rootClass,
}) => {
  const styles = Storyblok.arrayToMuiStyles(rootClass);

  const customIcon = customIcons.find(icon => icon.iconName === iconName);
  if (customIcon) {
    const Custom = customIcon.Component;
    return (
      <Custom
        className={styles.root}
        color={color}
        fontSize={fontSize}
        {...customIcon.props}
      />
    );
  }

  switch (iconName) {
    case 'facebook':
      return <FacebookSvg color={color} fontSize={fontSize} className={styles.root} />;
    case 'bbb':
      return <BBBSvg color={color} fontSize={fontSize} className={styles.root} />;
    case 'linkedin':
      return <LinkedinSvg color={color} fontSize={fontSize} className={styles.root} />;
    case 'google':
      return <GoogleSvg color={color} fontSize={fontSize} className={styles.root} />;
    default:
      return (
        <Icon
          className={styles.root}
          color={color}
          fontSize={fontSize}
        >
          {iconName}
        </Icon>
      );
  }
};

export default MuiIcon;

MuiIcon.propTypes = {
  /** any icon from https://material.io/resources/icons/?style=baseline */
  iconName: PropTypes.string.isRequired,
  /**
   * mui prop: 'inherit', 'primary', 'secondary', 'action', 'error', 'disabled'
   * The color of the component. It supports those theme colors that make sense for this component.
   * */
  color(props, propName, componentName) {
    const validProp = ['inherit', 'primary', 'secondary', 'action', 'error', 'disabled'];
    return muiStringProp(props, propName, componentName, validProp);
  },
  /**
   * mui prop: 'default', 'small', 'inherit', 'large'
   * The fontSize applied to the icon. Defaults to 24px, but can be configure to inherit font size.
   * */
  fontSize(props, propName, componentName) {
    const validProp = ['default', 'small', 'inherit', 'large'];
    return muiStringProp(props, propName, componentName, validProp);
  },
  /**
   * storyblok multiselect of css classes
   * Override or extend the styles applied to the component
   * */
  rootClass: PropTypes.arrayOf(PropTypes.string),
};

MuiIcon.defaultProps = {
  rootClass: [],
  color: 'secondary',
  fontSize: 'default',
};
