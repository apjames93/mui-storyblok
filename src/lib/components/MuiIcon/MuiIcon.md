MuiIcon example:
[View All MUI Icons](https://material.io/resources/icons/?style=baseline)

```js
  <div styles={{ display: 'flex', flexDirection: 'row' }}>
    <MuiIcon
      iconName='home'
      color='primary'
    />
    <MuiIcon
      iconName='menu'
      color='secondary'
    />
    <MuiIcon
      iconName='close'
      color='action'
    />
    <MuiIcon
      iconName='accessibility'
      color='inherit'
    />
    <MuiIcon
      iconName='done'
      color='error'
    />

      <MuiIcon
      iconName='done'
      color='error'
      fontSize="small"
    />
  </div>
```

Social Icons for MuiIcon examples:
* iconName = 'facebook'
* iconName = 'bbb'
* iconName = 'google'
* iconName = 'linkedin'
* iconName = 'youtube'
* iconName = 'instagram'
* iconName = 'twitter'
* iconName = 'yelp'
* iconName = 'github'

```js
  <div styles={{ display: 'flex', flexDirection: 'row' }}>
    <MuiIcon
      iconName='facebook'
      color='primary'
      fontSize='large'
    />
    <MuiIcon
      iconName='bbb'
      color='secondary'
      fontSize='large'
    />
    <MuiIcon
      iconName='google'
      color='action'
      fontSize='large'
    />
    <MuiIcon
      iconName='linkedin'
      color='inherit'
      fontSize='large'
      rootClass={['margin-left: 10px', 'margin-right: 10px']}
    />
    <MuiIcon
      iconName='youtube'
      color='inherit'
      fontSize='large'
    />
    <MuiIcon
      iconName='instagram'
      color='secondary'
      fontSize='large'
    />
    <MuiIcon
      iconName='twitter'
      color='primary'
      fontSize='large'
    />
    <MuiIcon
      iconName='yelp'
      color='secondary'
      fontSize='large'
    />
    <MuiIcon
      iconName='github'
      color='inherit'
      fontSize='large'
    />
  </div>
```