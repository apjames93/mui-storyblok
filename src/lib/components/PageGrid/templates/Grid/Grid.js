import React, { Suspense } from 'react';
import { Grid as MuiGrid, Box } from '@material-ui/core';
import PropTypes from 'prop-types';
import Storyblok from 'lib/utils/Storyblok';
import { muiStringProp, dimensionProp } from 'lib/utils/customProps';
import { renderComponentsWithBridge } from 'lib/utils/customComponents';
import GridItem from 'lib/components/PageGrid/templates/GridItem/GridItem';

const Grid = ({
  alignContent,
  alignItems,
  rootClass,
  direction,
  justify,
  wrap,
  spacing,
  content,
  components,
  style,
  dataBlokC,
  dataBlokUid,
  storyblokClass,
  height,
  backgroundImageUrl,
  transition,
}) => {
  // let gridClass = useFullBorderedGridStyles({ borderColor: 'primary.main' });
  // gridClass = window?.Storyblok?.inEditor ? gridClass : {};
  const gridClass = window?.Storyblok?.inEditor ? {
    borderStyle: 'solid',
    borderColor: '#FF2020',
    borderWidth: '.1em',
  } : {};

  let heroClass = {
    ...style,
    height,
  };

  if (backgroundImageUrl) {
    heroClass = {
      ...heroClass,
      ...{
        backgroundImage: `url(${backgroundImageUrl})`,
        backgroundPosition: 'center',
        backgroundRepeat: 'no-repeat',
        backgroundSize: 'cover',
        position: 'relative',
        padding: 0,
        margin: 0,
      },
    };
  }

  const styles = Storyblok.arrayToMuiStyles(rootClass, { ...heroClass });

  return (
    <Box width={{ xs: '100%' }}>
      <MuiGrid
        container
        alignContent={alignContent}
        alignItems={alignItems}
        direction={direction}
        justify={justify}
        wrap={wrap}
        spacing={Number(spacing)}
        data-blok-c={dataBlokC}
        data-blok-uid={dataBlokUid}
        className={`${styles.root} ${storyblokClass}`}
        style={gridClass}
      >
        {!content.length && <Box minHeight={200} width={{ xs: '100%' }} />}
        {content.length > 0 && content.map((component, key) => (
          <Suspense fallback={<Box width={{ xs: '100%' }} key={key} />}>
            {renderComponentsWithBridge({ ...{ GridItem }, ...components }, {
              ...component,
              components,
              key,
            }, key)}
          </Suspense>
        ))}
      </MuiGrid>
    </Box>
  );
};

export default Grid;

Grid.propTypes = {
  /** url for background img */
  backgroundImageUrl: PropTypes.string,
  /** height of the container */
  height(props, propName, componentName) {
    return dimensionProp(props, propName, componentName);
  },
  /**
   * stroyblok multiselect of css classes
   * Mui Override or extend the styles applied to the component. See CSS API below for more details.
   */
  rootClass: PropTypes.arrayOf(PropTypes.string),
  /**
   * mui prop: 'stretch', 'center', 'flex-start', 'flex-end', 'space-between', 'space-around'
   * Defines the align-content style property. It's applied for all screen sizes.
   */
  alignContent(props, propName, componentName) {
    const validProp = ['stretch', 'center', 'flex-start', 'flex-end', 'space-between', 'space-around'];
    return muiStringProp(props, propName, componentName, validProp);
  },
  /**
   * mui prop: 'flex-start' , 'center' , 'flex-end' , 'stretch' , 'baseline'
   * Defines the align-items style property. It's applied for all screen sizes.
   */
  alignItems(props, propName, componentName) {
    const validProp = ['flex-start', 'center', 'flex-end', 'stretch', 'baseline'];
    return muiStringProp(props, propName, componentName, validProp);
  },
  /*
  * mui prop: 'row', 'row-reverse', 'column', 'column-reverse'
  * Defines the flex-direction style property. It is applied for all screen sizes.
  */
  direction(props, propName, componentName) {
    const validProp = ['row', 'row-reverse', 'column', 'column-reverse'];
    return muiStringProp(props, propName, componentName, validProp);
  },
  /**
   * mui prop:
   * 'flex-start', 'center', 'flex-end', 'space-between', 'space-around', 'space-evenly'
   * Defines the justify-content style property. It is applied for all screen sizes.
   */
  justify(props, propName, componentName) {
    const validProp = ['flex-start', 'center', 'flex-end', 'space-between', 'space-around', 'space-evenly'];
    return muiStringProp(props, propName, componentName, validProp);
  },
  /**
   * mui prop:  0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
   * Defines the space between the type item
   component. It can only be used on a type container component.
   */
  spacing: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * mui prop: 'nowrap', 'wrap', 'wrap-reverse'
   * Defines the flex-wrap style property. It's applied for all screen sizes.
   */
  wrap(props, propName, componentName) {
    const validProp = ['nowrap', 'wrap', 'wrap-reverse'];
    return muiStringProp(props, propName, componentName, validProp);
  },
  /**
   * Content passed to render
   */
  content: PropTypes.arrayOf(PropTypes.shape({
    component: PropTypes.string.isRequired,
  })).isRequired,

  /** Transition desired to apply on grid item. */
  transition: PropTypes.string,
  /** style obj to pass to grid */
  style: PropTypes.shape(),

  /** storyblok prop for when in editor to allow click bridge */
  dataBlokC: PropTypes.string,
  /** storyblok prop for when in editor to allow click bridge */
  dataBlokUid: PropTypes.string,
  /** storyblok prop for when in editor to allow click bridge */
  storyblokClass: PropTypes.string,
  /**  components to render in the GridItem */
  components: PropTypes.shape(),
};

Grid.defaultProps = {
  alignContent: 'stretch',
  alignItems: 'center',
  rootClass: [],
  direction: 'row',
  justify: 'center',
  spacing: '0',
  wrap: 'wrap',
  style: {},
  dataBlokC: '',
  dataBlokUid: '',
  storyblokClass: '',
  components: {},
  height: '100%',
  backgroundImageUrl: '',
  transition: '',
};
