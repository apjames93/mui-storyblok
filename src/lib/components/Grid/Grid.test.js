import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Grid from './Grid';
import MuiIcon from '../MuiIcon/MuiIcon';

function setup() {
  const props = {
    _editable: '<!--#storyblok#{"name": "Grid", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
    components: {
      MuiIcon,
    },
    content: [{
      _editable: '<!--#storyblok#{"name": "GridItem", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
      component: 'GridItem',
      content: [{
        _editable: '<!--#storyblok#{"name": "MuiIcon", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
        component: 'MuiIcon',
        iconName: 'android',
      }],
    }],
  };
  const comp = shallow(<Grid {...props} />);
  return { comp, props };
}

describe('<Grid />', () => {
  it('renders Grid', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  test('snapshot', () => {
    const { props } = setup();

    const tree = renderer.create(<Grid {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
