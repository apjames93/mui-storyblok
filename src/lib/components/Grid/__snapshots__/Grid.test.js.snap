// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> snapshot 1`] = `
<div
  className="MuiGrid-root makeStyles-root-3 MuiGrid storyblok--outline MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  data-blok-c="{\\"name\\": \\"Grid\\", \\"space\\": \\"48408\\", \\"uid\\": \\"947be9f0-47c3-4315-a95a-550f0c560eb5\\", \\"id\\": \\"307934\\"}"
>
  <div
    className="MuiGrid-root makeStyles-root-4 GridItem storyblok--outline MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    data-blok-c="{\\"name\\": \\"GridItem\\", \\"space\\": \\"48408\\", \\"uid\\": \\"947be9f0-47c3-4315-a95a-550f0c560eb5\\", \\"id\\": \\"307934\\"}"
  >
    <span
      aria-hidden={true}
      className="material-icons MuiIcon-root makeStyles-root-5 MuiIcon-colorSecondary"
    >
      android
    </span>
  </div>
</div>
`;
